geowave/src/main/scala/geotrellis/spark/io/geowave/GeowaveLayerReader.scala

```diff
@@ -0,0 +1,243 @@
+    * @param  ranges   The ranges in degree of longitude and latitude associated with the chosen zoom level
+    */
+  def computeMetadata(
+    adapter: RasterDataAdapter,
+    ranges: Array[Double]
+  ): (TileLayerMetadata[SpatialKey], Int, Int) = {
+    val adapterId = adapter.getAdapterId
+
+    val bbox = bboxMap.getOrElse(adapterId, throw new Exception(s"Unknown Adapter Id $adapterId"))
+
+    val minX = bbox.getMinX
+    val minY = bbox.getMinY
+    val maxX = bbox.getMaxX
+    val maxY = bbox.getMaxY
+    val minCol = (minX / ranges(0)).toInt
+    val minRow = (minY / ranges(1)).toInt
+    val maxCol = (maxX / ranges(0)).toInt
+    val maxRow = (maxY / ranges(1)).toInt
+
+    val extent = Extent(
+      minRow * ranges(1),
+      minCol * ranges(0),
+      (maxRow + 1) * ranges(1),
+      (maxCol + 1) * ranges(0)
+    )
+
+    val layout = {
+      val tileSize = adapter.getTileSize
+      val tileLayout = TileLayout(maxCol - minCol + 1, maxRow - minRow + 1, tileSize, tileSize)
+      LayoutDefinition(extent, tileLayout)
+    }
+
+    val cellType = {
+      val geom = (new GeometryFactory).createPoint(new Coordinate((minX + maxX) / 2.0, (minY + maxY) / 2.0))
+      val queryOptions = new QueryOptions(adapter, index)
+      val query = new IndexOnlySpatialQuery(geom)
+      val config = computeConfiguration
+      GeoWaveInputFormat.setQuery(config, query)
+      GeoWaveInputFormat.setQueryOptions(config, queryOptions)
+
+      val gc = sc.newAPIHadoopRDD(
+        config,
+        classOf[GeoWaveInputFormat[GridCoverage2D]],
+        classOf[GeoWaveInputKey],
+        classOf[GridCoverage2D])
+        .map({ case (_, gc) => gc })
+        .collect.head
+
+      GridCoverage2DConverters.getCellType(gc)
+    }
+
+    val bounds = KeyBounds(
+      SpatialKey(0, 0),
+      SpatialKey(maxRow - minRow, maxCol - minCol) // sic
+    )
+
+    (TileLayerMetadata(cellType, layout, extent, LatLng, bounds), maxCol, minRow)
+  }
+
+  /**
+    * Read particular rasters out of the GeoWave database.  The
+    * particular rasters to read are given by the result running the
+    * provided LayerQuery.
+    *
+    * @param  id               The LayerId specifying the name and zoom level to query
+    * @param  rasterQuery      Produces a list of rasters to read
+    * @param  numPartitions    The number of Spark partitions to use
+    * @param  filterIndexOnly  ?
+    */
+  def read[
+    K: AvroRecordCodec: Boundable: JsonFormat: ClassTag,
+    V: AvroRecordCodec: ClassTag,
+    M: JsonFormat: GetComponent[?, Bounds[K]]
+  ](id: LayerId, rasterQuery: LayerQuery[K, M], numPartitions: Int, filterIndexOnly: Boolean) = {
+    import GeoWaveLayerReader._
+
+    if (!attributeStore.layerExists(id))
+      throw new LayerNotFoundError(id)
+    if (implicitly[ClassTag[K]].toString != "geotrellis.spark.SpatialKey")
+      throw new Exception("Unsupported Key Type")
+
+    /* XXX numPartitions, and filterIndexOnly are currently ignored. */
```

Comment:
In GeoWave, is it always 1 index == 1 tile? If so, then you're not really ignoring filter index only, it's just by default doing either true or false behavior. `filterIndexOnly` is for when the index has more than one tile stored in the values, so that when we are updating tile entries in a backend we can update the tile information while not losing the surrounding tile info.